### PR TITLE
feat: Add `allowed_instance_types` to `instance_requirements`

### LINKED ISSUE #265: Unable to define allowed_instance_types when defining a launch_template
## Description

Please provide a clear and concise description of the issue you are encountering, and a reproduction of your configuration (see the `examples/*` directory for references that you can copy+paste and tailor to match your configs if you are unable to copy your exact configuration). The reproduction MUST be executable by running `terraform init && terraform apply` without any further changes.

If your request is for a new feature, please use the `Feature request` template.

- [x] ✋ I have searched the open/closed issues and my issue is not listed.

## ⚠️ Note

Before you submit an issue, please perform the following first:

1. Remove the local `.terraform` directory (! ONLY if state is stored remotely, which hopefully you are following that best practice!): `rm -rf .terraform/`
2. Re-initialize the project root to pull down modules: `terraform init`
3. Re-attempt your terraform plan or apply and check if the issue still persists

## Versions

- Module version [Required]: 7.4.1

- Terraform version: v1.8.0

- Provider version(s):
  - hashicorp/aws: 5.45.0

## Reproduction Code [Required]

```
module "asg" {
  source = "terraform-aws-modules/autoscaling/aws"
  version = "7.4.1"

  name                   = "example"
  autoscaling_group_tags = local.cluster_tags
  update_default_version = true

  create_launch_template = true
  image_id               = local.image_id
  launch_template_name   = local.cluster_name
  user_data              = base64encode(local.user_data)

  security_groups = local.security_groups
  vpc_zone_identifier = data.aws_subnets.private_subnets.ids

  default_cooldown      = 10
  max_size              = 1
  min_size              = 1

  health_check_type         = "EC2"
  health_check_grace_period = 600

  use_mixed_instances_policy = true
  instance_requirements = {
    allowed_instance_types = ["t3.micro", "t2.micro"]
    burstable_performance  = "included"

    memory_mib = {
      min = 256
      max = 8388608
    }

    vcpu_count = {
      min = 1
      max = 8096
    }
  }
}
```

```
terraform apply
```

## Expected behavior

When running `terraform apply` my expectation is that the list provided for `instance_requirements.allowed_instance_types` gets populated within the EC2 launch template when it get created.

I expected it to be populated in the same way that it does when using `mixed_instances_policy.override`. However, when using this method, the `allowed_instance_types` get added to the autoscaling group instead of the launch template (this is expected).

Code like shown here:
```
  mixed_instances_policy = {
    override = [
      {
        instance_requirements = {
          allowed_instance_types = ["t3.micro", "t2.micro"]
        }
      }
    ]
  }
```

## Actual behavior

The launch template is created but the list provided for `instance_requirements.allowed_instance_types` doesn't get populated into the created launch template.

## Additional context

I at the source and found that the problem is due to a missing line of code.

When looking at [main.tf:769](https://github.com/terraform-aws-modules/terraform-aws-autoscaling/blob/master/main.tf#L769), I found that `allowed_instance_types` is utilized within the `override` block. However, I would expect to find the same logic on [main.tf:182](https://github.com/terraform-aws-modules/terraform-aws-autoscaling/blob/master/main.tf#L182), but don't. I have not verified this theory, yet, but adding the following line at [main.tf:182](https://github.com/terraform-aws-modules/terraform-aws-autoscaling/blob/master/main.tf#L182) should remedy the issue I'm experiencing:
```
allowed_instance_types = try(instance_requirements.value.allowed_instance_types, null)
```

### PR DESCRIPTION
**Pull request by baolsen**
## Description

The [aws_launch_template](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/launch_template) resource supports an `allowed_instance_types` field in the `instance_requirements` block.

> [allowed_instance_types](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/launch_template#allowed_instance_types) - (Optional) List of instance types to apply your specified attributes against. All other instance types are ignored, even if they match your specified attributes. You can use strings with one or more wild cards, represented by an asterisk (*), to allow an instance type, size, or generation. The following are examples: m5.8xlarge, c5*.*, m5a.*, r*, *3*. For example, if you specify c5*, you are allowing the entire C5 instance family, which includes all C5a and C5n instance types. If you specify m5a.*, you are allowing all the M5a instance types, but not the M5n instance types. Maximum of 400 entries in the list; each entry is limited to 30 characters. Default is all instance types.

> NOTE: If you specify allowed_instance_types, you can't specify excluded_instance_types.

This terraform module does already accept the `excluded_instance_types` field but not the `allowed_instance_types` field documented on the resource.

## Motivation and Context

Closes #265 

## Breaking Changes

No - I've included an update the `excluded_instance_types` to ensure compatibility when either field is present.

## How Has This Been Tested?

- [x] I have updated at least one of the `examples/*` to demonstrate and validate my change(s)
- [x] I have tested and validated these changes using one or more of the provided `examples/*` projects

- [x] I have executed `pre-commit run -a` on my pull request

### CODE REVIEW

**Comment by baolsen** [comment]
I've tested this a bit locally. 

Terraform plan with the following combinations:
1
```
allowed_instance_types = ["c6in.xlarge"]

=> works
```

2
```
excluded_instance_types = ["c6in.xlarge"]

=> works
```

3 
```
<none>

=> works
```

4
```
allowed_instance_types = ["c6in.xlarge"]
excluded_instance_types = ["c6in.xlarge"]

=> 
Planning failed. Terraform encountered an error while generating this plan.
Error: Conflicting configuration arguments
"instance_requirements.0.allowed_instance_types": conflicts with instance_requirements.0.excluded_instance_types
Error: Conflicting configuration arguments
"instance_requirements.0.excluded_instance_types": conflicts with instance_requirements.0.allowed_instance_types
```

**Comment by baolsen** [comment]
I see there is an issue template that for some reason I didn't see when creating this :)
Will update the description accordingly

**Comment by baolsen** [comment]
@antonbabenko , ready for review

**Comment by bryantbiggs** [comment]
the min required AWS provider version throughout will need to be raised to `5.45` to support this

**Comment by baolsen** [comment]
Thank you @bryantbiggs 

I had to go with `5.46` here to cater for the min version needed on `https://registry.terraform.io/modules/terraform-aws-modules/alb/aws/9.9.0`.

Basically, I can't use `5.45` here without also constraining the ALB module in the example to `<9.9.0`

Happy to do that if you think its worth it

File: README.md [hunk 1]
```diff
@@ -225,19 +225,45 @@ Note: the default behavior of the module is to create an autoscaling group and l
 
 - [Complete](https://github.com/terraform-aws-modules/terraform-aws-autoscaling/tree/master/examples/complete) - Creates several variations of resources for autoscaling groups and launch templates.
 
+## Contributing
```
Comment (bryantbiggs) [inline README.md:228]: please remove this

**Comment by bryantbiggs** [comment]
The min version is to declare the modules minimum requirements - it doesn't mean you can pin to this exact version in an example because the use of other resources/modules in an example may require a higher version. It is just a lower bound pin to declare the minimum requirements for the module itself; Terraform will use this when resolving the version requirements in conjunction with other modules and user defined version requirements

**Comment by baolsen** [comment]
@bryantbiggs thank you, should be sorted now

**Review by bryantbiggs — APPROVED** [review]

**Comment by antonbabenko** [comment]
This PR is included in version 7.5.0 :tada:

**Pull request merged**